lib/src/rules/avoid_type_to_string.dart

```diff
@@ -0,0 +1,189 @@
+    // Also delegates to visitArgumentList.
+    registry.addMethodInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  final TypeSystem typeSystem;
+  final InterfaceType typeType;
+
+  // Null if there is no logical `this` in the given context.
+  InterfaceType thisType;
+
+  _Visitor(this.rule, this.typeSystem, this.typeType);
+
+  @override
+  void visitClassDeclaration(ClassDeclaration node) {
+    thisType = node.declaredElement.thisType;
+  }
+
+  @override
+  void visitMixinDeclaration(MixinDeclaration node) {
+    thisType = node.declaredElement.thisType;
+  }
+
+  @override
+  void visitExtensionDeclaration(ExtensionDeclaration node) {
+    // Might not be InterfaceType. Ex: visiting an extension on a dynamic type.
+    thisType = (node.declaredElement.extendedType is InterfaceType)
+        ? node.declaredElement.extendedType as InterfaceType
+        : null;
+  }
+
+  @override
+  void visitFunctionExpressionInvocation(FunctionExpressionInvocation node) {
+    visitArgumentList(node.argumentList);
+  }
+
+  @override
+  void visitInstanceCreationExpression(InstanceCreationExpression node) {
+    visitArgumentList(node.argumentList);
+  }
+
+  @override
+  void visitRedirectingConstructorInvocation(
+      RedirectingConstructorInvocation node) {
+    visitArgumentList(node.argumentList);
+  }
+
+  @override
+  void visitSuperConstructorInvocation(SuperConstructorInvocation node) {
+    visitArgumentList(node.argumentList);
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    visitArgumentList(node.argumentList);
+
+    final targetType = (node.realTarget?.staticType is InterfaceType)
+        ? node.realTarget.staticType as InterfaceType
+        : thisType;
+    _reportIfToStringOnCoreTypeClass(node, targetType, node.methodName);
+  }
+
+  @override
+  void visitArgumentList(ArgumentList node) {
+    node.arguments.forEach(_validateArgument);
+  }
+
+  void _validateArgument(Expression expression) {
+    if (expression is PropertyAccess) {
+      final targetType = (expression.realTarget?.staticType is InterfaceType)
+          ? expression.realTarget?.staticType as InterfaceType
+          : thisType;
+      _reportIfToStringOnCoreTypeClass(
+          expression, targetType, expression.propertyName);
+    } else if (expression is PrefixedIdentifier) {
+      final prefixType = expression.prefix.staticType;
+      if (prefixType is InterfaceType) {
+        _reportIfToStringOnCoreTypeClass(
+            expression, prefixType, expression.identifier);
+      }
+    } else if (expression is SimpleIdentifier) {
+      _reportIfToStringOnCoreTypeClass(expression, thisType, expression);
+    }
+  }
+
+  void _reportIfToStringOnCoreTypeClass(AstNode offendingNode,
+      InterfaceType targetType, SimpleIdentifier methodIdentifier) {
+    if (_isToStringOnCoreTypeClass(targetType, methodIdentifier)) {
+      rule.reportLint(offendingNode);
```

Comment:
We generally associate diagnostics with a single token where possible because it improves the UX. In this case, it would be better to report the lint on `methodIdentifier` and drop the `offendingNode` parameter.

Reply:
Makes sense. 👍